picard/util/pipe.py

```diff
@@ -0,0 +1,287 @@
+# -*- coding: utf-8 -*-
+#
+# Picard, the next-generation MusicBrainz tagger
+#
+# Copyright (C) 2022 skelly37
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from typing import Optional
+
+from picard import PICARD_APP_ID
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+
+
+if IS_WIN:
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+
+class PipeError(Exception):
+    pass
```

Comment:
What about doing something like:

```python
class PipeError(Exception):
    MESSAGE = None

    def __init__(self, *messages):
        if self.MESSAGE:
            self.messages = (self.MESSAGE, ) + tuple(messages)
        else:
            self.messages = tuple(messages)

    def __str__(self):
        messages_str = "\n  ".join(str(m) for m in self.messages)
        if not messages_str:
            messages_str = "unknown"
        return f"ERROR: {messages_str}"


class PipeErrorInvalidArgs(PipeError):
    MESSAGE = "Pipe() args argument has to be iterable"

```

So following code:

```python
try:
    raise PipeError()
except PipeError as exc:
    print(exc)

try:
    raise PipeError("errmsg")
except PipeError as exc:
    print(exc)

try:
    raise PipeError("errmsg1", "errmsg2", 3, 4)
except PipeError as exc:
    print(exc)
    print(exc.messages)


try:
    raise PipeErrorInvalidArgs("extramsg")
except PipeError as exc:
    print(exc)

```

Would output:

```
ERROR: unknown
ERROR: errmsg
ERROR: errmsg1
  errmsg2
  3
  4
('errmsg1', 'errmsg2', 3, 4)
ERROR: Pipe() args argument has to be iterable
  extramsg
```

It would reduce code redundancy and unify error messages.